### PR TITLE
Assignment9+10 multithreading

### PR DESCRIPTION
**Pull request by j0shh3ss**
Bhopss152
Brendan Hopkins

### CODE REVIEW

**Review by AnnikaStachowiak — COMMENTED** [review]
Instructions

    Implement PingProcess' public Task<PingResult> RunTaskAsync(string hostNameOrAddress) ✔
        First implement public void RunTaskAsync_Success() test method to test PingProcess.RunTaskAsync() using "localhost". ✔
        Do NOT use async/await in this implementation. ✔
    Implement PingProcess' async public Task<PingResult> RunAsync(string hostNameOrAddress) ✔
        First implement the public void RunAsync_UsingTaskReturn_Success() test method to test PingProcess.RunAsync() using "localhost" without using async/await. ✔
        Also implement the async public Task RunAsync_UsingTpl_Success() test method to test PingProcess.RunAsync() using "localhost" but this time DO using async/await. ✔
    Add support for an optional cancellation token to the PingProcess.RunAsync() signature. ✔ Inside the PingProcess.RunAsync() invoke the token's ThrowIfCancellationRequested() method so an exception is thrown. ✔ Test that, when cancelled from the test method, the exception thrown is an AggregateException ✔ with a TaskCanceledException inner exception ✔ using the test methods RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping ✔and RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException ✔ respectively.
    Complete/fix AND test async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default) which executes ping for an array of hostNameOrAddresses (which can all be "localhost") in parallel, adding synchronization if needed. ✔ NOTE:
        The order of the items in the stdOutput is irrelevant and expected to be intermingled.
        StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
    Implement AND test public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token) using Task.Factory.StartNew() and invoking RunProcessInternal with a TaskCreation value of TaskCreationOptions.LongRunning and a TaskScheduler value of TaskScheduler.Current. Returning a Task<PingResult> is also okay. NOTE: This method does NOT use Task.Run.

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.✔
    Place all shared project items into a Directory.Build.targets file.❌
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
    and enabled .NET analyzers for both projects ✔
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

RungLongRunningAsync test and StringBuilderAppendLine test do not pass. Langversion is not set. Just a couple of small issues, but looks good otherwise.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -53,79 +54,157 @@ public void Run_CaptureStdOutput_Success()
...
+   // var pingProcess = new PingProcess("google.com");
 
+    // Act
+    //var pingResult = await pingProcess.RunLongRunningAsync_UsingTpl();
+
+    // Assert
+    //AssertValidPingOutput(pingResult);
+//}
+
+    private StringBuilder _stringBuilder = new StringBuilder();
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/PingProcessTests.cs:185]: image

File: Assignment/Assignment/PingProcess.cs [hunk 2]
```diff
@@ -25,39 +25,101 @@ public PingResult Run(string hostNameOrAddress)
         return new PingResult( process.ExitCode, stringBuilder?.ToString());
     }
 
-    public Task<PingResult> RunTaskAsync(string hostNameOrAddress)
+    public async Task<PingResult> RunTaskAsync(string hostNameOrAddress)
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:28]: This should _NOT_ be using async/await

File: Assignment/Assignment/PingProcess.cs [hunk 3]
```diff
@@ -25,39 +25,101 @@ public PingResult Run(string hostNameOrAddress)
...
     {
-        throw new NotImplementedException();
+        StartInfo.Arguments = hostNameOrAddress;
+        StringBuilder? stringBuilder = null;
+        void updateStdOutput(string? line) =>
+            (stringBuilder ??= new StringBuilder()).AppendLine(line);
+        Process process = RunProcessInternal(StartInfo, updateStdOutput, default, default);
+        var processExited = new TaskCompletionSource<bool>();
+        process.Exited += (sender, args) => processExited.TrySetResult(true);
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:36]: You shouldn't need to deal with process.Exited here, just call RunProcessInternal and return the result of that.

File: Assignment/Assignment/PingProcess.cs [hunk 4]
```diff
@@ -25,39 +25,101 @@ public PingResult Run(string hostNameOrAddress)
...
 
-    async public Task<PingResult> RunAsync(params string[] hostNameOrAddresses)
+
+    async public Task<PingResult> RunAsync(string hostNameOrAddress, CancellationToken cancellationToken = default)
     {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        StartInfo.Arguments = hostNameOrAddress;
         StringBuilder? stringBuilder = null;
-        ParallelQuery<Task<int>>? all = hostNameOrAddresses.AsParallel().Select(async item =>
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:44]: You should keep the ParallelQuery here

**Pull request closed without merging**